src/connection-loop.ts

```diff
@@ -0,0 +1,229 @@
+import {resolver} from './resolver.js';
+import {sleep} from './sleep.js';
+
+const DEBOUNCE_DELAY = 10;
+
+const MIN_DELAY = 30;
+const MAX_DELAY = 60_000;
+
+const MAX_CONNECTIONS = 3;
+
+type Duration = {duration: number; ok: boolean};
+type Durations = Duration[];
+
+export interface ConnectionLoopDelegate {
+  closed(): boolean;
+  invokeExecute(): unknown;
+  debounceDelay?: number;
+  maxConnections?: number;
+  printTime(): string;
+  watchdogTimer?: number;
+}
+
+export class ConnectionLoop {
+  private _pendingResolver = resolver<void>();
+  private _currentResolver = resolver<void>();
+  private _pushCounter = 0;
+
+  private readonly _delegate: ConnectionLoopDelegate;
+  private _lastSendTime = 0;
+
+  constructor(delegate: ConnectionLoopDelegate) {
+    this._delegate = delegate;
+    this.run();
+  }
+
+  async execute(): Promise<void> {
```

Comment:
Nit: I think I would call this something like `send()`. To me *execute* implies code is running or something.